service/forward_service.cc

```diff
@@ -201,20 +201,16 @@ future<query::forward_result> forward_service::dispatch_to_shards(
 ) {
     _stats.requests_dispatched_to_own_shards += 1;
 
-    auto result = co_await container().map_reduce0(
+    return container().map_reduce0(
         [req, tr_info] (auto& fs) {
             return fs.execute_on_this_shard(req, tr_info);
         },
-        std::optional<query::forward_result>(),
-        [&req] (std::optional<query::forward_result> partial, query::forward_result mapped) -> std::optional<query::forward_result>{
-            if (partial) {
-                mapped.merge(*partial, req.reduction_types);
-            }
-            return {mapped};
+        query::forward_result(),
+        [reduction_types = req.reduction_types] (query::forward_result partial, query::forward_result mapped) -> query::forward_result {
+            mapped.merge(partial, reduction_types);
+            return mapped;
```

Comment:
LGTM

A tiny improvement could be if merge would return *this as forward_result&
and then the reducer could just `return mapped.merge(partial, reduction_types)`

Reply:
Would you like to see it done in this pull request?

Reply:
Yes, please, it would be nice to have while we're at it and it will enjoy the CI testing as well.